File: board/lib/rules/gear_ratio.c
#include <stddef.h>
#include "gear_ratio.h"

typedef struct {
	float gear;
	float ratio;
} GearRatio;

/* RATIOS in rules/instant/gear_ratio.py */
static const GearRatio RATIOS[] = {
	{2.0f, 179.25f}, {4.0f, 108.69f}, {5.0f, 87.02f}, {6.0f, 67.20f},
	{7.0f, 52.27f}, {8.0f, 41.37f}, {9.0f, 31.69f}, {10.0f, 24.82f},
	{11.0f, 19.37f}, {12.0f, 15.24f},
};
#define GEARS (sizeof(RATIOS) / sizeof(RATIOS[0]))

static bool in_table(float gear)
{
	size_t i;

	for(i = 0; i < GEARS; i++) {
		if(RATIOS[i].gear == gear) {
			return true;
		}
	}
	return false;
}

/* The gear whose ratio is closest to `ratio`, compared as a proportion. */
static float nearest_gear(float ratio)
{
	size_t i, best = 0;
	float distance, best_distance = 0.0f;

	for(i = 0; i < GEARS; i++) {
		distance = RATIOS[i].ratio / ratio;
		if(ratio / RATIOS[i].ratio > distance) {
			distance = ratio / RATIOS[i].ratio;
		}
		if(i == 0 || distance < best_distance) {
			best = i;
			best_distance = distance;
		}
	}
	return RATIOS[best].gear;
}

bool gear_ratio_hits(float engine_speed, float wheel_speed, float current_gear,
	float selected_gear, float clutch_slip, float min_speed)
{
	/* mid shift or with the clutch open there is no fixed ratio */
	if(wheel_speed < min_speed || !in_table(current_gear)
		|| current_gear != selected_gear || clutch_slip != 0.0f) {
		return false;
	}
	/* the closed clutch turns the engine with the wheels, so a stopped engine
	 * contradicts the speed above */
	if(engine_speed <= 0.0f) {
		return true;
	}
	return nearest_gear(engine_speed / wheel_speed) != current_gear;
}

File: rules/instant/gear_ratio.py
"""Flag the engine and wheel speeds not matching the gear the transmission reports.

Each gear turns the engine a set number of times per km/h. The gears are spaced 1.28
apart, so a measured ratio picks out one of them, and it should be the reported one.
No tolerance is needed for that, only the table.
"""

from __future__ import annotations

import numpy as np

from preprocess.features.signal_state import SIGNALS

# The median of engine_speed / wheel_speed in each gear, over 150,291 evaluations
# with the clutch closed. Gears 1 and 3 are too rare in the data to place.
RATIOS = {2: 179.25, 4: 108.69, 5: 87.02, 6: 67.20, 7: 52.27,
          8: 41.37, 9: 31.69, 10: 24.82, 11: 19.37, 12: 15.24}

def nearest_gear(ratio: np.ndarray, ratios: dict = RATIOS) -> np.ndarray:
    """The gear whose ratio is closest to each of `ratio`, compared as a proportion."""
    ratio = np.asarray(ratio)
    ratio = ratio.astype(np.result_type(ratio, np.float32))[..., None]
    gears = np.array(list(ratios))
    table = np.array(list(ratios.values()), dtype=ratio.dtype)
    return gears[np.maximum(table / ratio, ratio / table).argmin(axis=-1)]


def hits(raw: np.ndarray, min_speed: float, ratios: dict = RATIOS) -> np.ndarray:
    """True where the speeds pick out a gear other than the reported one.

    Below `min_speed` the wheel speed is too coarse for the ratio, as in shaft_ratio.
    """
    engine = raw[:, SIGNALS.index("engine_speed")]
    wheel = raw[:, SIGNALS.index("wheel_speed")]
    gear = raw[:, SIGNALS.index("current_gear")]
    selected = raw[:, SIGNALS.index("selected_gear")]
    slip = raw[:, SIGNALS.index("clutch_slip")]
    # mid shift or with the clutch open there is no fixed ratio
    fixed = ((wheel >= min_speed) & np.isin(gear, list(ratios)) & (gear == selected)
             & (slip == 0))
    turning = fixed & (engine > 0)
    ratio = np.divide(engine, wheel, out=np.ones_like(wheel), where=turning)
    # the closed clutch turns the engine with the wheels, so a stopped engine
    # contradicts the speed above
    return (fixed & (engine <= 0)) | (turning & (nearest_gear(ratio, ratios) != gear))
